src/domain.rs

```diff
@@ -181,6 +188,93 @@ macro_rules! unique_domain {
+/// Type.
+///
+/// This is useful since it allows you to store the [`HazardPointer`]s aquired from this domain, since
+/// the Family type can now be named.
+///
+/// ```rust
+/// # struct DataStructure;
+/// # use haphazard::{HazardPointer, AtomicPtr, static_unique_domain};
+/// static_unique_domain!(static LOCAL: Domain<Local>);
+///
+/// struct ContainsHazardPointers<'domain> {
+///     haz_ptr: HazardPointer<'domain, Local>,
+///     val: AtomicPtr<DataStructure, Local>,
+/// }
+///
+/// impl ContainsHazardPointers<'_> {
+///     fn read(&mut self) -> Option<&DataStructure> {
+///         self.val.safe_load(&mut self.haz_ptr)
+///     }
+/// }
+/// # fn main() {}
+/// ```
+///
+/// # Notes
+///
+/// This macro cannot be used in a function scope or impl block. This macro (at least currently)
+/// requires the ability to define a module, and therefore must be placed in module scope. This
+/// may be relaxed in the future, but hopefully shouldn't be exteneded.
+///
+/// This also restricts the ability to define a struct or module with the same name as the static
+/// variable, or the family.
+#[macro_export]
+macro_rules! static_unique_domain {
+    ($v:vis static $domain:ident: Domain<$family:ident>) => {
+        #[allow(non_snake_case)]
+        mod $domain {
+            struct UniqueFamily;
+            // Safety: nowhere else can construct an instance of UniqueFamily to pass to Domain::new.
+            unsafe impl $crate::Singleton for UniqueFamily {}
+            pub struct $family {
+                // Technically, since we are in an inner module, this private member may actually
+                // be enough, it might not need to be a private type.
+                _inner: UniqueFamily,
+            }
+            // Safety: $family can only be constructed by first constructing an instance of
+            // UniqueFamily, which is itself a Singlton.
+            unsafe impl $crate::Singleton for $family {}
+            pub static $domain: $crate::Domain<$family> = $crate::Domain::new(&$family {
+                _inner: UniqueFamily,
+            });
+        }
+        $v use $domain::$family;
+        $v use $domain::$domain;
+    };
+}
+
+/// This item is defined purely to run compile_fail doc tests. It is marked as `doc(hidden)`, to
+/// avoid polluting the documentation.
+///
+/// ```rust,compile_fail
+/// # struct DataStructure;
+/// # use haphazard::{HazardPointer, AtomicPtr, static_unique_domain};
+/// static_unique_domain!(static LOCAL: Domain<Local>);
+/// static BROKEN: Domain<Local> = Domain::new(&Local {
+///     _inner: LOCAL::UniqueFamily,
+/// });
+/// # fn main() {}
+/// ```
+///
+/// ```rust
+/// # struct DataStructure;
+/// # use haphazard::{HazardPointer, AtomicPtr, static_unique_domain};
+/// static_unique_domain!(static LOCAL: Domain<Local>);
+/// static_unique_domain!(static LOCAL2: Domain<Local2>);
+/// # fn main() {}
+/// ```
+#[doc(hidden)]
+fn compile_fail_tests() {}
```

Comment:
I may be misremembering, but I think this has to be `pub` for its doc tests to actually be run. I believe this was a change in a semi-recent version of Rust. Or maybe I have it backwards that doctests _used_ to only be run for `pub` items, but are now run for everything. Just good to double-check.

Reply:
I know that my version of Rust runs this doc-test, but I think I will add pub to ensure these tests run in most versions of rust.